### PR TITLE
[12.0] [ADD] fieldservice_route

### PR DESCRIPTION
**Pull request by nikul-serpentcs**
Issue: #380

### CODE REVIEW

File: fieldservice_route/models/fsm_order.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            if not fsm_route:
+                fsm_route_obj.create({
+                    'person_id': res.person_id.id,
+                    'scheduled_date': res.scheduled_date_start,
+                    'date': datetime.now()
+                })
+        for fsm_route_id in fsm_route_obj.search([('order_ids', 'in', res.id)]):
+            if fsm_route_id.days:
+                fsm_route_id.days = False
+        return res
```
Comment (max3903) [inline fieldservice_route/models/fsm_order.py:48]: @nikul-serpentcs You need to extract this logic into his own method so you can call it from `create()` and `write()`

File: fieldservice_route/models/fsm_order.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+from odoo import api, models
+
+
+class FSMOrder(models.Model):
+    _inherit = 'fsm.order'
+
+    @api.model
+    def create(self, vals):
+        res = super(FSMOrder, self).create(vals)
+        fsm_route_obj = self.env['fsm.route']
```
Comment (max3903) [inline fieldservice_route/models/fsm_order.py]:
```suggestion
        fsm_route_obj = self.env['fsm.route.dayroute']
```

File: fieldservice_route/models/fsm_route_dayroute.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class FSMRouteDayRoute(models.Model):
+    _name = 'fsm.route.dayroute'
+
+    @api.multi
+    @api.depends('route_id')
+    def _compute_order_count(self):
+        for rec in self:
+            rec.order_count = 0
+            if rec.route_id and rec.route_id.order_ids:
```
Comment (max3903) [inline fieldservice_route/models/fsm_route_dayroute.py]:
```suggestion
            if rec.order_ids:
```

File: fieldservice_route/models/fsm_route_dayroute.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+                rec.order_count = len(rec.route_id.order_ids)
```
Comment (max3903) [inline fieldservice_route/models/fsm_route_dayroute.py]:
```suggestion
                rec.order_count = len(rec.order_ids)
```

File: fieldservice_route/models/fsm_route_dayroute.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    last_location_id = fields.Many2one('fsm.location', string='Last Location')
+    date_start_planned = fields.Datetime(string='Planned Start Time')
+    start_location_id = fields.Many2one('fsm.location', string='Start Location')
+    end_location_id = fields.Many2one('fsm.location', string='End Location')
+    work_time = fields.Float(string='Time before overtime (in hours)',
+                             default=8.0)
+    max_allow_time = fields.Float(string="Maximal Allowable Time (in hours)",
+                                  default=10.0)
+    order_count = fields.Integer(string='Number of Orders',
+                                 compute=_compute_order_count)
```
Comment (max3903) [inline fieldservice_route/models/fsm_route_dayroute.py:44]: @nikul-serpentcs Please move the code from here:
https://github.com/OCA/field-service/blob/12.0/fieldservice/models/fsm_route.py
to this module.

You forgot the `name` and `order_ids` fields, the method and the constraint.

**Comment by murtuzasaleh** [comment]
#440

File: fieldservice_route/models/fsm_route_dayroute.py (hunk 4 above)
Comment (murtuzasaleh) [reply fieldservice_route/models/fsm_route_dayroute.py]: Done

File: fieldservice_route/models/fsm_route_dayroute.py (hunk 5 above)
Comment (murtuzasaleh) [reply fieldservice_route/models/fsm_route_dayroute.py:44]: Done

File: fieldservice_route/models/fsm_route_dayroute.py (hunk 3 above)
Comment (murtuzasaleh) [reply fieldservice_route/models/fsm_route_dayroute.py]: Done

File: fieldservice_route/models/fsm_order.py (hunk 2 above)
Comment (murtuzasaleh) [reply fieldservice_route/models/fsm_order.py]: Done

File: fieldservice_route/models/fsm_route_dayroute.py (hunk 5 above)
Comment (max3903) [reply fieldservice_route/models/fsm_route_dayroute.py:44]: @murtuzasaleh Still missing the method:
https://github.com/OCA/field-service/blob/12.0/fieldservice/models/fsm_route.py#L31

and the constraint:
https://github.com/OCA/field-service/blob/12.0/fieldservice/models/fsm_route.py#L25

File: fieldservice_route/security/ir.model.access.csv [hunk 6]
```diff
@@ -0,0 +1,7 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_fsm_route_user,fsm.route.user,model_fsm_route,fieldservice.group_fsm_user,1,1,0,0
+access_fsm_route_dispatcher,fsm.route.dispatcher,model_fsm_route,fieldservice.group_fsm_dispatcher,1,1,1,1
+access_fsm_route_day_user,fsm.route.day.user,model_fsm_route_day,fieldservice.group_fsm_user,1,1,0,0
+access_fsm_route_day_dispatcher,fsm.route.day.dispatcher,model_fsm_route_day,fieldservice.group_fsm_dispatcher,1,1,1,1
+access_fsm_route_dayroute_user,fsm.route.dayroute.user,model_fsm_route_dayroute,fieldservice.group_fsm_user,1,1,0,0
+access_fsm_route_dayroute_dispatcher,fsm.route.dayroute.dispatcher,model_fsm_route_dayroute,fieldservice.group_fsm_dispatcher,1,1,1,1
```
Comment (max3903) [inline fieldservice_route/security/ir.model.access.csv]: Missing newline.

File: fieldservice_route/views/menu.xml [hunk 7]
```diff
@@ -0,0 +1,36 @@
+<odoo>
+
+    <!-- Routes -->
+    <menuitem id="menu_fsm_route_config"
```
Comment (max3903) [inline fieldservice_route/views/menu.xml]:
```suggestion
    <menuitem id="menu_fsm_route_operations"
```

File: fieldservice_route/views/menu.xml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+              parent="menu_fsm_route_config"
+              sequence="10"
+              groups="base.group_no_one"/>
+
+    <menuitem id="menu_fsm_route"
+              name="Routes"
+              action="action_fsm_route"
+              parent="menu_fsm_route_config"
+              sequence="20"
+              groups="base.group_no_one"/>
```
Comment (max3903) [inline fieldservice_route/views/menu.xml]: This menuitem should be under `Master Data` and you can remove the groups attribute.

File: fieldservice_route/views/menu.xml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+    <menuitem id="menu_fsm_route_day"
+              name="Route Days"
+              action="action_fsm_route_day"
+              parent="menu_fsm_route_config"
+              sequence="30"
+              groups="base.group_no_one"/>
```
Comment (max3903) [inline fieldservice_route/views/menu.xml]: This one should be in Configuration and you can remove groups.

File: fieldservice_route/views/menu.xml [hunk 10]
```diff
@@ -0,0 +1,36 @@
+<odoo>
+
+    <!-- Routes -->
+    <menuitem id="menu_fsm_route_config"
+              name="Routes"
+              parent="fieldservice.operations"/>
```
Comment (max3903) [inline fieldservice_route/views/menu.xml]: You can remove this menu item,

File: fieldservice_route/views/menu.xml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+    <menuitem id="menu_fsm_route_dayroute"
+              name="Day Routes"
+              action="action_fsm_route_dayroute"
+              parent="menu_fsm_route_config"
+              sequence="10"
+              groups="base.group_no_one"/>
```
Comment (max3903) [inline fieldservice_route/views/menu.xml]: you can remove the groups attribute.

File: fieldservice_route/views/menu.xml [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    <!-- Routes -->
+    <menuitem id="menu_fsm_route_config"
+              name="Routes"
+              parent="fieldservice.operations"/>
+
+    <menuitem id="menu_fsm_route_dayroute"
+              name="Day Routes"
+              action="action_fsm_route_dayroute"
+              parent="menu_fsm_route_config"
```
Comment (max3903) [inline fieldservice_route/views/menu.xml]:
```suggestion
              parent="fieldservice.operations"
```

File: fieldservice_route/models/fsm_route_dayroute.py (hunk 5 above)
Comment (murtuzasaleh) [reply fieldservice_route/models/fsm_route_dayroute.py:44]: Done

File: fieldservice_route/views/menu.xml (hunk 8 above)
Comment (murtuzasaleh) [reply fieldservice_route/views/menu.xml]: Done

File: fieldservice_route/views/menu.xml (hunk 9 above)
Comment (murtuzasaleh) [reply fieldservice_route/views/menu.xml]: Done

File: fieldservice_route/views/menu.xml (hunk 10 above)
Comment (murtuzasaleh) [reply fieldservice_route/views/menu.xml]: Done

File: fieldservice_route/views/menu.xml (hunk 11 above)
Comment (murtuzasaleh) [reply fieldservice_route/views/menu.xml]: Done

File: fieldservice_route/views/menu.xml (hunk 12 above)
Comment (murtuzasaleh) [reply fieldservice_route/views/menu.xml]: Done

File: fieldservice_route/models/fsm_order.py (hunk 1 above)
Comment (murtuzasaleh) [reply fieldservice_route/models/fsm_order.py:48]: Done

Comment (max3903) [reply fieldservice_route/models/fsm_order.py:48]: @murtuzasaleh Add a new method `_manage_fsm_route()` to include this logic and then call this method from `create()` and `write()`

**Comment by murtuzasaleh** [comment]
@max3903 @dreispt Can you please review the PR.

**Comment by brian10048** [comment]
I did a quick functional test on the runbot. A few comments below. Otherwise it matches the spec and is looking good to me.

Show the day route on the FSM order (maybe under times in schedule details, or under assigned to field?)

Show the day route's stage as a header.

I expected when a day route was created the route field would be automatically set based on a route matching the assigned worker and day of week.  This can be implemented later.

**Comment by murtuzasaleh** [comment]
> I did a quick functional test on the runbot. A few comments below. Otherwise it matches the spec and is looking good to me.
> 
> Show the day route on the FSM order (maybe under times in schedule details, or under assigned to field?)
> 
> Show the day route's stage as a header.
> 
> I expected when a day route was created the route field would be automatically set based on a route matching the assigned worker and day of week. This can be implemented later.

Added `Day route` field in view after the `Scheduled End` field (Schedule Details).

File: fieldservice_route/readme/DESCRIPTION.rst [hunk 13]
```diff
@@ -0,0 +1,4 @@
+A Field Service route is the sorted list of orders that one worker/vehicle will perform during a day.
+
+This module allows you to manage your Field Service routes: when assigning an order to a worker and scheduling it, a route is automatically created for that worker and day.
+
```
Comment (max3903) [inline fieldservice_route/readme/DESCRIPTION.rst:4]:
```suggestion
```

File: fieldservice_route/readme/USAGE.rst [hunk 14]
```diff
@@ -0,0 +1,7 @@
+To use this module, you need to:
+
+* Go to Field Service
+* Create or select an order
+* Assign it to a worker and schedule it
+* Go to Field Service > Dashboard > Day Routes. A new record has been created.
+
```
Comment (max3903) [inline fieldservice_route/readme/USAGE.rst:7]:
```suggestion
```

File: fieldservice_route/models/fsm_order.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+                    })
+                    vals.update({'dayroute_id': fsm_dayroute.id})
+                else:
+                    vals.update({'dayroute_id': False})
+
+    @api.model
+    def create(self, vals):
+        res = super(FSMOrder, self).create(vals)
+        if res.person_id and res.scheduled_date_start:
+            res._manage_fsm_route()
```
Comment (max3903) [inline fieldservice_route/models/fsm_order.py]:
```suggestion
            res._manage_fsm_route(vals)
```

File: fieldservice_route/models/fsm_order.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+                if fsm_route:
+                    vals.update({'dayroute_id': fsm_route.id})
+                elif not fsm_route and person_id and scheduled_date_start:
+                    fsm_dayroute = fsm_route_obj.create({
+                        'person_id': person_id,
+                        'date': scheduled_date_start,
+                    })
+                    vals.update({'dayroute_id': fsm_dayroute.id})
+                else:
+                    vals.update({'dayroute_id': False})
```
Comment (max3903) [inline fieldservice_route/models/fsm_order.py]: @nikul-serpentcs If the order is moved to another day route (date or worker is changed) or removed (date or worker are unset), the day route needs to be archived if there is no more order on it.

Try this scenario:
* Create an order with a date and worker = Worker 1 --> SR001 is created
* Edit the order to change the worker = Worker 2 --> SR002 is created, but SR001 is still active. SR001 should be archived.
* Edit the order to  change the worker = Worker 1 --> SR001 is active and has one order. SR002 is archived and is empty.

**Comment by max3903** [comment]
@nikul-serpentcs Can you:

- [x] add a many2one field "Route" (route_id) on fsm.location to fsm.route
- [x] add a related field "Route" on fsm.order to fsm_location_id.route_id

File: fieldservice_route/views/fsm_route_dayroute.xml [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+                <field name="date"/>
+            </tree>
+        </field>
+    </record>
+
+    <record id="fsm_route_dayroute_form" model="ir.ui.view">
+        <field name="name">fsm.route.dayroute.form</field>
+        <field name="model">fsm.route.dayroute</field>
+        <field name="arch" type="xml">
+            <form string="Day Route">
```
Comment (max3903) [inline fieldservice_route/views/fsm_route_dayroute.xml:19]:
```suggestion
            <form string="Day Route">
                <header>
                    <field name="stage_id"
                           widget="statusbar"
                           options="{'fold_field': 'fold'}"
                           clickable="True"
                           domain="[('stage_type', '=', 'route')]"/>
                </header>
```

File: fieldservice_route/tests/test_fsm_order.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+        self.fsm_stage_obj = self.env['fsm.stage']
+        self.fsm_order_obj = self.env['fsm.order']
+        self.test_person = self.env.ref("fieldservice.test_person")
+        self.test_location = self.env.ref("fieldservice.test_location")
+
+    def test_create_day_route(self):
+        self.fsm_order_obj.create({'location_id': self.test_location.id,
+                                   'person_id': self.test_person.id,
+                                   'scheduled_date_start': datetime.now(),
+                                   })
```
Comment (max3903) [inline fieldservice_route/tests/test_fsm_order.py]: @nikul-serpentcs Can you add the tests to:

- [ ] check that the route of the order is the same as the route of the location
- [ ] check that the order is added to a dayroute with the same person, date and route

Comment (murtuzasaleh) [reply fieldservice_route/tests/test_fsm_order.py]: Done

**Review by max3903 — APPROVED** [review]

**Review by murtuzasaleh — APPROVED** [review]

**Pull request merged**